File: ios/Releaf/Data/Notebook/PageRepository.swift
/*
 * PageRepository.swift
 *
 * Single-page CRUD + FTS-backed search. Mirror of Android's
 * `PageRepository.kt`. Shape matches `NotepadRepository` for consistent
 * VM-side logic: observe, findById, create, save, softDelete,
 * undoSoftDelete, plus sync-worker helpers.
 */

import Foundation
import GRDB

public struct PageSearchHit: Codable, FetchableRecord, Equatable, Identifiable, Sendable {
    public let id: String
    public let title: String?
    public let notes: String
    public let updatedAt: String
    public let notebookTitle: String
    public let chapterTitle: String

    public enum CodingKeys: String, CodingKey {
        case id
        case title
        case notes
        case updatedAt     = "updated_at"
        case notebookTitle = "notebook_title"
        case chapterTitle  = "chapter_title"
    }

    public init(
        id: String,
        title: String?,
        notes: String,
        updatedAt: String,
        notebookTitle: String,
        chapterTitle: String
    ) {
        self.id = id
        self.title = title
        self.notes = notes
        self.updatedAt = updatedAt
        self.notebookTitle = notebookTitle
        self.chapterTitle = chapterTitle
    }
}

public final class PageRepository: @unchecked Sendable {

    private let dbQueue: DatabaseQueue

    public init(database: ReleafDatabase = .shared) {
        self.dbQueue = database.dbQueue
    }

    // MARK: - Observation

    public func observeForChapter(chapterId: String) -> AsyncThrowingStream<[PageEntity], Error> {
        let observation = ValueObservation.tracking { db in
            try PageEntity
                .filter(sql: "chapter_id = ? AND deleted_at IS NULL", arguments: [chapterId])
                .order(Column("position").asc, Column("created_at").asc)
                .fetchAll(db)
        }
        return bridge(observation.values(in: dbQueue))
    }

    /// All live pages across a notebook, newest-edited first. Joins
    /// through `chapters` so tombstoned chapters also hide their pages.
    public func observeForNotebook(notebookId: String) -> AsyncThrowingStream<[PageEntity], Error> {
        let observation = ValueObservation.tracking { db in
            try PageEntity.fetchAll(db, sql: """
                SELECT p.* FROM pages p
                JOIN chapters c ON c.id = p.chapter_id
                WHERE c.notebook_id = ?
                  AND p.deleted_at IS NULL
                  AND c.deleted_at IS NULL
                ORDER BY p.updated_at DESC
                """, arguments: [notebookId])
        }
        return bridge(observation.values(in: dbQueue))
    }

    public func observeById(_ id: String) -> AsyncThrowingStream<PageEntity?, Error> {
        let observation = ValueObservation.tracking { db in
            try PageEntity
                .filter(sql: "id = ? AND deleted_at IS NULL", arguments: [id])
                .fetchOne(db)
        }
        return bridge(observation.values(in: dbQueue))
    }

    public func findById(_ id: String) async throws -> PageEntity? {
        try await dbQueue.read { db in
            try PageEntity
                .filter(sql: "id = ?", arguments: [id])
                .fetchOne(db)
        }
    }

    // MARK: - FTS search

    /// Full-text search across every live page. Uses the shared
    /// `FtsQuery` sanitizer so empty/noise queries short-circuit to an
    /// empty snapshot rather than raising an FTS5 MATCH error.
    public func searchAll(rawQuery: String) -> AsyncThrowingStream<[PageEntity], Error> {
        guard let match = FtsQuery.build(rawQuery) else {
            return AsyncThrowingStream { c in
                c.yield([])
                c.finish()
            }
        }
        let observation = ValueObservation.tracking { db in
            try PageEntity.fetchAll(db, sql: """
                SELECT p.* FROM pages p
                JOIN fts_page_notes fts ON fts.page_id = p.id
                WHERE p.deleted_at IS NULL
                  AND fts_page_notes MATCH ?
                ORDER BY rank
                """, arguments: [match])
        }
        return bridge(observation.values(in: dbQueue))
    }

    /// Full-text search across every live page, plus the notebook/chapter
    /// labels needed to make tab-level results understandable at a glance.
    public func searchAllWithContext(rawQuery: String) -> AsyncThrowingStream<[PageSearchHit], Error> {
        guard let match = FtsQuery.build(rawQuery) else {
            return AsyncThrowingStream { c in
                c.yield([])
                c.finish()
            }
        }
        let observation = ValueObservation.tracking { db in
            try PageSearchHit.fetchAll(db, sql: """
                SELECT
                    p.id,
                    p.title,
                    p.notes,
                    p.updated_at,
                    c.title AS chapter_title,
                    n.title AS notebook_title
                FROM pages p
                JOIN chapters c ON c.id = p.chapter_id
                JOIN notebooks n ON n.id = c.notebook_id
                JOIN fts_page_notes fts ON fts.page_id = p.id
                WHERE p.deleted_at IS NULL
                  AND c.deleted_at IS NULL
                  AND n.deleted_at IS NULL
                  AND fts_page_notes MATCH ?
                ORDER BY rank
                """, arguments: [match])
        }
        return bridge(observation.values(in: dbQueue))
    }

    public func searchInNotebook(notebookId: String, rawQuery: String) -> AsyncThrowingStream<[PageEntity], Error> {
        guard let match = FtsQuery.build(rawQuery) else {
            return AsyncThrowingStream { c in
                c.yield([])
                c.finish()
            }
        }
        let observation = ValueObservation.tracking { db in
            try PageEntity.fetchAll(db, sql: """
                SELECT p.* FROM pages p
                JOIN chapters c ON c.id = p.chapter_id
                JOIN fts_page_notes fts ON fts.page_id = p.id
                WHERE c.notebook_id = ?
                  AND p.deleted_at IS NULL
                  AND c.deleted_at IS NULL
                  AND fts_page_notes MATCH ?
                ORDER BY rank
                """, arguments: [notebookId, match])
        }
        return bridge(observation.values(in: dbQueue))
    }

    // MARK: - Create / Update

    @discardableResult
    public func createPage(chapterId: String, title: String? = nil, notes: String = "") async throws -> PageEntity {
        let now = IsoClock.nowIso()
        let entity = PageEntity(
            id:        Uuidv7.generate(),
            chapterId: chapterId,
            title:     title?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty,
            notes:     notes,
            createdAt: now,
            updatedAt: now,
            dirty:     true
        )
        try await dbQueue.write { db in try entity.insert(db) }
        return entity
    }

    public func savePage(_ entity: PageEntity) async throws {
        var row = entity
        row.title = entity.title?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
        row.updatedAt = IsoClock.nowIso()
        row.dirty = true
        // `let` snapshot before the @Sendable write closure — Swift 6
        // rejects capture-by-reference of mutable locals.
        let snapshot = row
        try await dbQueue.write { db in try snapshot.update(db) }
    }

    // MARK: - Soft delete

    public func softDeletePage(id: String) async throws {
        let now = IsoClock.nowIso()
        try await dbQueue.write { db in
            try db.execute(sql: """
                UPDATE pages
                SET deleted_at = ?, updated_at = ?, dirty = 1
                WHERE id = ?
                """, arguments: [now, now, id])
        }
    }

    public func undoSoftDeletePage(id: String) async throws {
        let now = IsoClock.nowIso()
        try await dbQueue.write { db in
            try db.execute(sql: """
                UPDATE pages
                SET deleted_at = NULL, updated_at = ?, dirty = 1
                WHERE id = ?
                """, arguments: [now, id])
        }
    }
}

// MARK: - Helpers

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

// MARK: - AsyncSequence bridge

private func bridge<S: AsyncSequence & Sendable>(_ sequence: S) -> AsyncThrowingStream<S.Element, Error>
where S.Element: Sendable {
    AsyncThrowingStream { continuation in
        let task = Task {
            do {
                for try await value in sequence {
                    continuation.yield(value)
                }
                continuation.finish()
            } catch {
                continuation.finish(throwing: error)
            }
        }
        continuation.onTermination = { _ in task.cancel() }
    }
}

File: apps/releaf/android/app/src/main/java/app/releaf/mobile/data/notebook/PageRepository.kt
/*
 * PageRepository.kt
 *
 * Single-page CRUD + FTS-backed search. Mirrors NotepadRepository in shape
 * (create / save / softDelete / undoSoftDelete / search) so the UI layer
 * uses the same mental model across both surfaces.
 */

package app.releaf.mobile.data.notebook

import androidx.room.RoomRawQuery
import app.releaf.mobile.data.common.FtsQuery
import app.releaf.mobile.data.common.IsoClock
import app.releaf.mobile.data.common.Uuidv7
import app.releaf.mobile.data.notepad.NotepadEntry
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.mapLatest

class PageRepository(
    private val pageDao: PageDao,
    /** See note on NotepadRepository — nullable for tests. */
    private val auditLogger: app.releaf.mobile.data.activity.AuditLogger? = null,
) {
    fun observeForChapter(chapterId: String): Flow<List<PageEntity>> =
        pageDao.observeForChapter(chapterId)

    fun observeForNotebook(notebookId: String): Flow<List<PageEntity>> =
        pageDao.observeForNotebook(notebookId)

    fun observeById(id: String): Flow<PageEntity?> = pageDao.observeById(id)

    /** page-count-per-notebook feed for the Notebooks tab rows. */
    fun observePageCountsByNotebook(): Flow<List<NotebookCountRow>> =
        pageDao.observePageCountsByNotebook()

    /** page-count-per-chapter feed for the notebook detail Chapter rows. */
    fun observePageCountsByChapter(): Flow<List<ChapterCountRow>> =
        pageDao.observePageCountsByChapter()

    /** Pages that carry at least one todo (open or done), newest-edit
     *  first. The caller is responsible for parsing the per-row
     *  `todos` JSON and filtering to open items. Drives the Open-todos
     *  modal on the library header. */
    fun observePagesWithTodos(): Flow<List<PageTodosRow>> =
        pageDao.observePagesWithTodos()

    suspend fun findById(id: String): PageEntity? = pageDao.findById(id)

    /**
     * Global FTS search across all live pages. Empty or all-noise queries
     * short-circuit to an empty flow so we don't pass a bad MATCH expression
     * into SQLite.
     *
     * `RoomRawQuery` is required by the DAO because `fts_page_notes` isn't
     * a Room-managed entity; see `PageDao.searchAllActive` for the rationale.
     */
    @OptIn(ExperimentalCoroutinesApi::class)
    fun searchAll(rawQuery: String): Flow<List<PageEntity>> {
        val match = FtsQuery.build(rawQuery) ?: return flowOf(emptyList())
        return pageDao.observeAllActive()
            .mapLatest { pageDao.searchAllActive(buildSearchAllActiveQuery(match)) }
            .distinctUntilChanged()
    }

    /** Global FTS search with notebook/chapter labels for UI context. */
    @OptIn(ExperimentalCoroutinesApi::class)
    fun searchAllWithContext(rawQuery: String): Flow<List<PageSearchHit>> {
        val match = FtsQuery.build(rawQuery) ?: return flowOf(emptyList())
        return pageDao.observeSearchScope()
            .mapLatest { pageDao.searchAllActiveWithContext(buildSearchAllWithContextQuery(match)) }
            .distinctUntilChanged()
    }

    /** Same as [searchAll] but scoped to a single notebook. */
    @OptIn(ExperimentalCoroutinesApi::class)
    fun searchInNotebook(notebookId: String, rawQuery: String): Flow<List<PageEntity>> {
        val match = FtsQuery.build(rawQuery) ?: return flowOf(emptyList())
        return pageDao.observeForNotebook(notebookId)
            .mapLatest { pageDao.searchInNotebook(buildSearchInNotebookQuery(notebookId, match)) }
            .distinctUntilChanged()
    }

    private fun buildSearchAllActiveQuery(match: String): RoomRawQuery =
        RoomRawQuery(
            sql = """
                SELECT p.* FROM pages p
                JOIN fts_page_notes fts ON fts.page_id = p.id
                WHERE p.deleted_at IS NULL
                  AND fts_page_notes MATCH ?
                ORDER BY rank
            """.trimIndent(),
            onBindStatement = { it.bindText(1, match) },
        )

    private fun buildSearchAllWithContextQuery(match: String): RoomRawQuery =
        RoomRawQuery(
            sql = """
                SELECT
                    p.id AS id,
                    p.title AS title,
                    p.notes AS notes,
                    p.updated_at AS updatedAt,
                    n.title AS notebookTitle,
                    c.title AS chapterTitle
                FROM pages p
                JOIN chapters c ON c.id = p.chapter_id
                JOIN notebooks n ON n.id = c.notebook_id
                JOIN fts_page_notes fts ON fts.page_id = p.id
                WHERE p.deleted_at IS NULL
                  AND c.deleted_at IS NULL
                  AND n.deleted_at IS NULL
                  AND fts_page_notes MATCH ?
                ORDER BY rank
            """.trimIndent(),
            onBindStatement = { it.bindText(1, match) },
        )

    private fun buildSearchInNotebookQuery(notebookId: String, match: String): RoomRawQuery =
        RoomRawQuery(
            sql = """
                SELECT p.* FROM pages p
                JOIN chapters c ON c.id = p.chapter_id
                JOIN fts_page_notes fts ON fts.page_id = p.id
                WHERE c.notebook_id = ?
                  AND p.deleted_at IS NULL
                  AND c.deleted_at IS NULL
                  AND fts_page_notes MATCH ?
                ORDER BY rank
            """.trimIndent(),
            onBindStatement = {
                it.bindText(1, notebookId)
                it.bindText(2, match)
            },
        )

    suspend fun createPage(
        chapterId: String,
        title: String?,
        notes: String,
    ): PageEntity {
        val now = IsoClock.nowIso()
        val entity = PageEntity(
            id        = Uuidv7.generate(),
            chapterId = chapterId,
            title     = title?.trim()?.ifEmpty { null },
            notes     = notes,
            createdAt = now,
            updatedAt = now,
            dirty     = true,
        )
        pageDao.upsert(entity)
        auditLogger?.log(
            action     = app.releaf.mobile.data.activity.AuditAction.Created,
            entityType = app.releaf.mobile.data.activity.AuditEntity.Page,
            entityId   = entity.id,
            title      = entity.title ?: "Untitled page",
        )
        return entity
    }

    /**
     * Create a new page inside [chapterId] whose body is copied straight
     * from a notepad entry. All the JSON side channels (sub-pages,
     * attachments, todos, contacts, locations, strokes) are carried over
     * verbatim — both surfaces share the same JSON shapes, so no
     * conversion is needed. The caller is expected to soft-delete the
     * source notepad entry after this returns.
     */
    suspend fun createFromNotepadEntry(
        chapterId: String,
        source: NotepadEntry,
    ): PageEntity {
        val now = IsoClock.nowIso()
        val page = PageEntity(
            id            = Uuidv7.generate(),
            chapterId     = chapterId,
            projectId     = source.projectId,
            title         = source.title?.trim()?.ifEmpty { null },
            notes         = source.notes,
            contacts      = source.contacts,
            locations     = source.locations,
            todos         = source.todos,
            attachments   = source.attachments,
            sketchStrokes = source.sketchStrokes,
            subPages      = source.subPages,
            createdAt     = now,
            updatedAt     = now,
            dirty         = true,
        )
        pageDao.upsert(page)
        auditLogger?.log(
            action     = app.releaf.mobile.data.activity.AuditAction.Created,
            entityType = app.releaf.mobile.data.activity.AuditEntity.Page,
            entityId   = page.id,
            title      = page.title ?: "Untitled page",
        )
        return page
    }

    suspend fun savePage(entity: PageEntity) {
        val updated = entity.copy(
            title     = entity.title?.trim()?.ifEmpty { null },
            updatedAt = IsoClock.nowIso(),
            dirty     = true,
        )
        pageDao.upsert(updated)
        auditLogger?.log(
            action     = app.releaf.mobile.data.activity.AuditAction.Updated,
            entityType = app.releaf.mobile.data.activity.AuditEntity.Page,
            entityId   = updated.id,
            title      = updated.title ?: "Untitled page",
        )
    }

    suspend fun softDeletePage(id: String) {
        val snapshot = pageDao.findById(id)
        pageDao.softDelete(id = id, nowIso = IsoClock.nowIso())
        auditLogger?.log(
            action     = app.releaf.mobile.data.activity.AuditAction.Deleted,
            entityType = app.releaf.mobile.data.activity.AuditEntity.Page,
            entityId   = id,
            title      = snapshot?.title ?: "Untitled page",
        )
    }

    suspend fun undoSoftDeletePage(id: String) {
        pageDao.restore(id = id, nowIso = IsoClock.nowIso())
        val snapshot = pageDao.findById(id)
        auditLogger?.log(
            action     = app.releaf.mobile.data.activity.AuditAction.Restored,
            entityType = app.releaf.mobile.data.activity.AuditEntity.Page,
            entityId   = id,
            title      = snapshot?.title ?: "Untitled page",
        )
    }
}
